numba/core/typing/templates.py

```diff
@@ -885,12 +885,21 @@ def generic(self, args, kws):
         """
         Type the intrinsic by the arguments.
         """
-        from numba.core.extending_hardware import resolve_dispatcher_from_str
+        from numba.core.extending_hardware import (get_local_target,
+                                                   resolve_target_str,
+                                                   dispatcher_registry)
         from numba.core.imputils import builtin_registry
 
         cache_key = self.context, args, tuple(kws.items())
-        hwstr = self.metadata.get('hardware', 'cpu')
-        disp = resolve_dispatcher_from_str(hwstr)
+        hwstr = self.metadata.get('hardware', 'generic')
+        # Get the class for the target declared by the function
+        hw_clazz = resolve_target_str(hwstr)
+        # get the local target
+        target_hw = get_local_target(self.context)
+        # make sure the target_hw is in the MRO for hw_clazz else bail
+        if not target_hw.inherits_from(hw_clazz):
+            return None
```

Comment:
This is never exercised. It seems this condition will never trigger because the intrinsic will not be found in the wrong hardware. So, maybe this should trigger a ICE instead.

Reply:
I think this is true. The `BaseFunction.get_call_type` would be the way to resolve the intrinsic and it would filter out intrinsics that are invalid for the hardware target. Will make this an ICE.

Reply:
Patch is in 4e8786c